Adapting Daniel's code for contributor graphs

rando2 opened the pull request:
<!-- Hi there! Please use the template below as a guide for opening a pull request.
This template is designed to help the pull request process go smoother.
If anything doesn't make sense to you, please open the pull request anyway and leave entries blank as needed.
Thank you for your contribution!! -->

### Description of the proposed additions or changes
<!-- If the additions or changes aren't self explanatory feel free to leave a description or notes here. -->
This still needs some work, but it generates something resembling the graphs that Daniel made for the Deep Review. There are a lot of compatibility issues that I'm whack-a-mole-ing but hopefully in the homestretch now. One figure worked, the other two there is some ggplot error error I'll work on tomorrow.

Also just some silly stuff that needs to be cleaned up, like I didn't update the output file names and I can't remember a lot of people's github username to real name mappings, so some of those are very very wrong (like just people's usernames a second time!)

### Related issues
<!-- GitHub will link issues to pull requests if you list the related issue numbers like "#2, #17". -->


### Suggested reviewers (optional)
<!-- If there are particular people you think should review this pull request, please list their GitHub handle like "@rando2".-->


### Checklist
<!-- Please mark the following tasks as complete by putting an "x" in the box like "[x]".
If a particular item is not applicable to your pull request, please delete the item. -->
- [ ] Text is formatted so that each sentence is on its own line.
<!-- see why this is recommended here:https://github.com/greenelab/covid19-review/blob/627a5f45a2d1fcb2150d1466df63bb848f22a3d4/USAGE.md#manubot-markdown -->
- [ ] Pre-prints cited in this pull request have a GitHub issue opened so that they can be reviewed.

rando2 commented:
OK! This can now create two figures and the stats file about the commits. I was trying to get a third figure to show contribution by career stage, but I'm going to give up on that for now because it's non-trivial and doesn't seem like something we completely need. There is some issue with one of the date packages that is causing a lot of issues and made it so that third graph really isn't usable, regardless (I was hoping if I switched to career stage, it would be more readable, but I don't think I have time for the data manipulation required to do that!)

Here are the figs that worked for easy viewing:
![covid19-review-contribution-dot](https://user-images.githubusercontent.com/8975841/116442281-50072f00-a820-11eb-8df5-ff6f321a0fba.png)
![covid19-review-contribution-ridge](https://user-images.githubusercontent.com/8975841/116442288-51d0f280-a820-11eb-81ff-43964a171bb9.png)

cbrueffer commented:
Very cool! I think the colors in figure 1 may be a bit hard to interpret. It's looks like most people only deleted code, when in reality I suppose the number of additions/deletions is equal, and red it just plotted last? Maybe replacing each circle with a small bar plot may work, although no idea how hard to implement.

If this is either/or between the two plots, I think the first one is more expressive.

One small nit, maybe Ashwin N. Skelly (dot after N) for consistency.

rando2 commented:
Quick update, based on convo with Casey I changed to log scale -- this does look better but we would need to figure out how to toggle the y-axis scaling
![covid19-review-contribution-ridge](https://user-images.githubusercontent.com/8975841/116448235-e0e10900-a826-11eb-8582-b1e43f46de3f.png)

agitter commented:
> It's looks like most people only deleted code, when in reality I suppose the number of additions/deletions is equal, and red it just plotted last?

Yes, that's what's happening.  If contributors primarily edit existing lines, those edited lines count as both added and deleted in the commit.  One solution would be to plot additions only and caption the figure appropriately to describe that as edited and added lines.

I liked the ridge plot in the original deep review summary, but I like the dot plot more here.  Even with the log scaling, the ridge plot has a lot of lines that look mostly horizontal.  I think the dot plot more accurately reflects the activity on the manuscript.  There was a big burst in spring 2020 that established the foundation.  Then, the number of active contributors diminished over time, but several contributors remained active editing the text.

I'm curious what my massive deletion was in August 2020 😄

rando2 commented:
@cgreene @cbrueffer @agitter Here is a plot of just additions! 
![covid19-review-contribution-dot](https://user-images.githubusercontent.com/8975841/116731059-3c420100-a9b7-11eb-9c61-47a748e72bcb.png)

agitter commented:
I like it!

cbrueffer commented:
Nice! I agree that the number of deletions is probably not so interesting here, so leaving them out is a good option.

rando2 commented:
@agitter I think this code would now work to be integrated with our workflow. I added a `.sh` script to run the notebooks (adapted from Daniel's Deep Review notebooks) and also added these figures to the versioning. However, I have generally only used jupyter notebooks for teaching, so had to look up a lot of these solutions -- I'm happy to change anything that could be handled better.

The one trick here is that it won't work in the base covid19-review environment, so I think I will need to also modify `.github/workflows/update-external-resources.yaml` to run these scripts first in a separate environment, then switch to the normal external-resources environment to run the rest of the figure scripts and then version everything. [This page](https://sionwilliams.com/posts/2020-02-04_jupyter_ci/) does seem to outline a solution to a similar problem, but it seems like more of an overhaul than we probably want since this is a one-off at present.

ajlee21 commented on contrib-viz/01.contrib-git.ipynb:
```diff
@@ -0,0 +1,591 @@
+    }
+   ],
+   "source": [
+    "path = pathlib.Path('renamer.tsv')\n",
```
What are you manually renaming here?

rando2 replied:
```suggestion
    # Update any users whose GitHub names don't match their authorship names
    # Use information specified in tab-delim file renamer.tsv
    "path = pathlib.Path('renamer.tsv')\n",
```

ajlee21 commented on contrib-viz/01.contrib-git.ipynb:
```diff
@@ -0,0 +1,591 @@
+    "\n",
+    "        # Attempt to ignore relocated lines\n",
+    "        additions, deletions = (\n",
+    "            [x for x in additions if get_word_count(x) < 4 or x not in set(deletions)],\n",
```
Why less than 4?

rando2 replied:
Great question! This is mostly Daniel's code. I'll play around with it and figure it out!

agitter replied:
That heuristic originated in https://github.com/greenelab/meta-review/commit/3aa682608566d55daecf0051af6c1764d008d3c7  Daniel's commit message acknowledges it is not very precise:
> This method aims to ignore lines that were relocated in the manuscript.
It is not very advanced and probably doesn't work very well.

ajlee21 reviewed (approved):
Really cool analysis and visualization! I'm not sure if I can follow all the scaping calculations performed, but the everything seems reasonable.

I know you log-scaled the values, did you also log-scale the size of the circle marker too?

Just to clarify, is this a visual meant to help inform authorship order? If so, I wonder if there are additional metrics that would be useful like "number of reviews/comments" in addition to text edits?

agitter commented on contrib-viz/02.contrib-viz.ipynb:
```diff
@@ -0,0 +1,819 @@
+    "  ggplot2::geom_point(stroke=1) +\n",
+    "  #ggplot2::scale_y_discrete(limits=usage_df$username, position = 'top', labels=y_labels, name=NULL) #+\n",
+    "  ggplot2::scale_x_discrete(limits = year_month_breaks, labels=year_month_labeler, name=NULL) +\n",
+    "  ggplot2::scale_size_area(max_size = 3.8, name='word count') +\n",
```
Because we are only showing additions, we could remove the green circle and `added` and change `word count` to `words added`.  I agree with @ajlee21's suggestion to log-scale the size of the circle marker if we can.

rando2 replied:
These suggestions are done in the new version I'll push. But log scaling did not seem to work, it made the dots look very uniform. I agree it would be a more intuitive measure but can't seem to get it to visualize well.

agitter commented on contrib-viz/03.contrib-stats.ipynb:
```diff
@@ -0,0 +1,481 @@
+     ]
+    }
+   ],
+   "source": [
```
Do we need this block?  We already have different ways to track the number of authors that might conflict with this count.

rando2 replied:
Removed!

agitter commented on contrib-viz/03.contrib-stats.ipynb:
```diff
@@ -0,0 +1,481 @@
+      "  \"review_authors\": 52,\n",
+      "  \"github_stars\": 112,\n",
+      "  \"github_forks\": 75,\n",
+      "  \"merged_pull_requests\": 559,\n",
```
Do we want to include `merged_pull_requests` and the number of issues in the methods manuscript?  That's nice to show how active the project has been.

rando2 replied:
That's great idea! Added a sentence to the conclusion in #1033

agitter commented on contrib-viz/calc-contrib-stats.sh:
```diff
@@ -0,0 +1,13 @@
+jupyter nbconvert --to notebook --execute --ExecutePreprocessor.kernel_name=python3 --output output.ipynb contrib-viz/03.contrib-stats.ipynb
+
+# Remove temporary files
+rm contrib-viz/commits.tsv contrib-viz/contributors.tsv
```
Do the above commands also create `output.ipynb` that can be removed?  Or do we want to give the outputs unique names and store them?

rando2 replied:
We can remove it, I think. Everything we need is saved as a json or image file.

```suggestion
rm contrib-viz/commits.tsv contrib-viz/contributors.tsv output.ipynb
```

agitter commented on version-figures.sh:
```diff
@@ -10,6 +10,7 @@ echo "Using figure versions from commit $FIGURE_COMMIT_SHA"
 CSSE_STATS_JSON=csse/csse-stats.json
 EBM_STATS_JSON=ebmdatalab/ebmdatalab-stats.json
 CORD19_STATS_JSON=CORD-19/cord19-stats.json
+CONTRIB_STATS_JSON=contrib-viz/covid19-review-stats.json
```
I don't think this file needs to be modified yet.  We don't have the figure URLs in `covid19-review-stats.json` so they don't need to have the commit replaced in the URLs.  We could save that set of changes for another pull request that integrates this into the daily or weekly workflow.

rando2 replied:
OK that sounds great!

agitter reviewed (changes requested):
This is looking good.  I have a few small suggestions, but it is nearly ready to merge.  I reviewed the main analysis code lightly because it is derived from Daniel's code that we used successfully in another project.

rando2 commented:
@agitter here is the updated figure:
![image](https://user-images.githubusercontent.com/8975841/133165641-99868d82-3654-4108-9ebd-f4bc723bb436.png)

agitter reviewed (approved):
Look great!  Let's merge it and add it.

I will also work on adding this to a workflow at some point post-submission.

We could probably remove "username" from the y-axis label on the dot plot now that we have full names.  Let's leave it for now.  I can manually crop it out if we want to remove it for the DISCO submission.

agitter commented:
I merged this so I can test it within the manuscript.
